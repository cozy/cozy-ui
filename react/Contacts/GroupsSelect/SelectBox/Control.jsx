import PropTypes from 'prop-types'
import React from 'react'

import Button from '../../../Buttons'
import Icon from '../../../Icon'
import BottomSelectIcon from '../../../Icons/BottomSelect'
import { useI18n, useExtendI18n } from '../../../providers/I18n'
import { locales } from '../locales'

const Control = ({
  innerRef,
  innerProps,
  selectProps: { onControlClicked }
}) => {
  useExtendI18n(locales)
  const { t } = useI18n()

  return (
    <div ref={innerRef} {...innerProps}>
      <Button
        variant="secondary"
        size="small"
        onClick={onControlClicked}
        onTouchStart={onControlClicked}
        label={t('Contacts.GroupsSelect.manage')}
        startIcon={<Icon icon={BottomSelectIcon} width="12" />}
      />
    </div>
  )
}

Control.propTypes = {
  selectProps: PropTypes.object.isRequired
}

export default Control
