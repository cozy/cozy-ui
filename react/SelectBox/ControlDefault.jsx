import React from 'react'
import PropTypes from 'prop-types'

import { components } from 'react-select'

const ControlDefault = ({
  selectProps: { onControlClicked },
  innerProps,
  ...props
}) => {
  // onTouchStart is necessary on mobile
  // see https://github.com/JedWatson/react-select/issues/3806#issuecomment-541325710

  const enhancedInnerProps = {
    ...innerProps,
    onClick: onControlClicked,
    onTouchStart: onControlClicked
  }
  return <components.Control {...props} innerProps={enhancedInnerProps} />
}

ControlDefault.propTypes = {
  selectProps: PropTypes.object.isRequired
}

export default ControlDefault
