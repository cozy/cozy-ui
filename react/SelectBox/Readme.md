A select box for more advanced use cases, based on [`react-select` v2](https://deploy-preview-2289--react-select.netlify.com/home).

### Simple use case

You can use this exactly like you would use `react-select`. See the [official documentation](https://deploy-preview-2289--react-select.netlify.com/home) for more examples.

```
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<SelectBox options={options} />
```

### Custom control component

You can use the `reactSelectControl` HOC to turn an existing, custom component into the control button for the `Select`:

```
const reactSelectControl = require('../SelectBox').reactSelectControl;
const MyControl = (<button>toggle options</button>);
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<SelectBox 
  options={options}
  components={{
    Control: reactSelectControl(MyControl)
  }}
/>
```

### Options with checkboxes

There is a `CheckboxOption` component that can be used as an `Option`, and renders like this:

```
const CheckboxOption = require('../SelectBox').CheckboxOption;
const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' }
];

<SelectBox 
  options={options}
  isMulti
  components={{
    Option: CheckboxOption
  }}
/>
```
