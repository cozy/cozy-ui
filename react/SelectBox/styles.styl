@require '../../stylus/components/forms'
@require '../../stylus/settings/palette'
@require '../../stylus/settings/breakpoints'
@require '../../stylus/tools/mixins'

.select-control__input
    width 0
    height 0
    overflow hidden

.select--autowidth
    max-width 32rem

.select--disabled
    @extends $select--disabled

.select--fullwidth
    @extends $select--fullwidth
    padding-right 0

.select-option
    display flex
    flex-direction row
    justify-content flex-start
    align-items center
    padding rem(8)
    border-left rem(4) solid transparent
    color var(--charcoalGrey)
    transition all .2s ease-out
    white-space normal /* otherwise the tick can be wrongly placed in Firefox */

    &:hover:not(.select-option--disabled)
        background-color var(--paleGrey)
        cursor pointer

        .select-option__actions
          opacity 1

.select-option--focused:not(.select-option--disabled)
    background-color var(--paleGrey)

.select-option--selected
    background-color var(--paleGrey)
    border-left-color var(--primaryColor)

.select-option--disabled
    color var(--silver)
    cursor not-allowed

.select-option__checkbox
    margin-right rem(8)
    vertical-align top

.select-option__label
    flex-grow 1
    display flex
    justify-content space-between
    align-items center
    width 100%
    overflow hidden

.select-option__checkmark
    width rem(32)

.select-option__actions
    opacity 0
    white-space nowrap
    transition opacity .2s ease-out

    +medium-screen()
      opacity 1

.select__overlay
    // Covers all the app so that clicking outside ends up
    // not on an option (hence not selecting) but on the select
    // so that anything under does not get clicked.
    &::before
        content '\00a0'
        width 200vw
        height 200vh
        top -50vh
        left -50vh
        display block
        position fixed

.MenuList
    display flex
    flex-direction column

.Group
    overflow-y auto
    padding-top 0.25rem !important
    padding-bottom .25rem !important

.FixedGroup
    @extend .Group
    flex-shrink 0
    border-top silver 1px solid
