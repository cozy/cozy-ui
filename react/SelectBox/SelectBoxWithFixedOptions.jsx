import { components } from 'react-select'
import React from 'react'
import groupBy from 'lodash/groupBy'

import SelectBox from './SelectBox'
import { silver } from '../../stylus/settings/palette.json'

const optionPadding = 0.25

const fixedOptionsStyle = {
  borderBottomLeftRadius: '4px',
  borderBottomRightRadius: '4px',
  paddingTop: optionPadding + 'rem',
  paddingBottom: optionPadding + 'rem',
  borderTop: `1px solid ${silver}`,
  background: 'white',
  position: 'relative'
}

const FixedOptions = props => (
  <div style={fixedOptionsStyle}>{props.children}</div>
)

/*
  This is the hacky part, where fixed options vnodes are manually spliced
  from the options vnodes. Fixed options vnodes are returned along with
  the menuUi where fixed vnodes have been removed.

  We rely on react-select internal structure `menu > scroll captor > menu list > options`.
  This code should be updated if react-select's internal structure changes.

  https://github.com/cozy/cozy-ui/pull/501
*/
const extractFixed = children => {
  const menuList = children.props.children.props.children
    ? children.props.children.props.children.props
    : null
  if (!menuList) {
    return { fixed: [], children }
  }
  const options = menuList.children
  const { fixed, normal } = groupBy(
    options,
    vnode => (vnode.props.data.fixed === true ? 'fixed' : 'normal')
  )

  const newElement = normal.map(element => {
    return React.cloneElement(element)
  })
  return { fixed, newElement }
}

const MenuWithFixedOptions = props => {
  const { children } = props
  const { fixed, newElement: menuUi } = extractFixed(children)
  return (
    <div>
      <components.Menu {...props}>
        {menuUi}
        <FixedOptions>{fixed}</FixedOptions>
      </components.Menu>
    </div>
  )
}

const fixedMenuStyle = {
  zIndex: 10,
  overflow: 'hidden',
  // the space at the end is important otherwise the property gets autoprefixed (by what, I
  // don't know) and is no longer understood by the browser
  display: 'flex ',
  flexDirection: 'column'
}

const SelectBoxWithFixedOptions = props => (
  <SelectBox
    {...props}
    styles={{
      menu: base => ({
        ...base,
        ...fixedMenuStyle
      }),
      ...(props.styles || {})
    }}
    components={{
      Menu: MenuWithFixedOptions,
      ...(props.components || {})
    }}
  />
)

export default SelectBoxWithFixedOptions
