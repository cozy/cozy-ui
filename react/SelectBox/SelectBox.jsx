import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ReactSelect, { components } from 'react-select'
import classNames from 'classnames'

import { isIOSApp } from 'cozy-device-helper'

import styles from './styles.styl'
import Icon from '../Icon'
import { dodgerBlue, silver, coolGrey } from '../palette'
import withBreakpoints from '../helpers/withBreakpoints'

const heights = {
  tiny: '2rem',
  medium: '2.5rem',
  large: '3rem'
}

const customStyles = props => ({
  control: (base, state) => ({
    ...base,
    backgroundColor: 'transparent',
    border: state.isFocused
      ? `.0625rem solid ${dodgerBlue}`
      : `.0625rem solid ${silver}`,
    ':hover': {
      borderColor: coolGrey
    },
    borderRadius: '.1875rem',
    boxShadow: 'unset',
    height: heights[props.size],
    minHeight: heights[props.size]
  }),
  dropdownIndicator: base => ({
    ...base,
    marginRight: '.75rem',
    height: '.875rem',
    width: '.875rem',
    padding: 0
  }),
  indicatorSeparator: () => ({
    display: 'none'
  }),
  valueContainer: base => ({
    ...base,
    color: 'black'
  }),
  menu: base => ({
    ...base,
    zIndex: 10
  })
})

/**
 * Determines maxHeight property for menuList component. This value is computed
 * with a container element and the current SelectBox element.
 * The container element defines an element which the SelectBox should not
 * overflow.
 * @param  {object} container React reference to an element containing
 * the SelectBox
 * @param  {object} element    React reference to the ReactSelect element
 * @return {object}            A style object, with `menuList` property set.
 */
const computedMenuListHeightStyles = (container, selectElement) => {
  if (!(container && selectElement)) return {}

  const containerPaddingTop = container.style.paddingTop || '0px'
  const containerPaddingBottom = container.style.paddingBottom || '0px'

  return {
    menuList: base => {
      const basePaddings = (base.paddingTop || 0) + (base.paddingBottom || 0)
      const spaceLeft =
        container.getBoundingClientRect().bottom -
        selectElement.getBoundingClientRect().bottom -
        basePaddings
      return {
        ...base,
        // containerPaddingTop and containerPaddingBottom can be in `rem`, so
        // let's use calc()
        maxHeight: `calc(${spaceLeft}px - ${containerPaddingTop} - ${containerPaddingBottom})`
      }
    }
  }
}

const DropdownIndicator = props => {
  return (
    <components.DropdownIndicator {...props}>
      <Icon
        icon={props.selectProps.menuIsOpen ? 'top' : 'bottom'}
        color={coolGrey}
        width="20"
        height="16"
      />
    </components.DropdownIndicator>
  )
}

const reactSelectControl = CustomControl => ({
  innerProps,
  innerRef,
  children
}) => (
  <div {...innerProps} ref={innerRef}>
    {CustomControl}
    <div className={styles['select-control__input']}>{children}</div>
  </div>
)

export const withPrefix = (reactSelectCx, className) => {
  // react-select implement a classnames function https://git.io/fhT8S
  // it's not the same as classnames library (https://www.npmjs.com/package/classnames)
  // 1st parameter is bound by react-select with prefix https://git.io/fhkIj
  // 2nd parameter is cssKey. We don't need it so it's set to null
  // 3rd parameter is used to add prefix but we don't want to stick with
  //   webpack className so we add space (' ') in front of it
  const classNameWithPrefix = reactSelectCx(null, { [` ${className}`]: true })

  // When we don't use classNamePrefix cx return '' so we verify to send
  // className
  return classNameWithPrefix === '' ? className : classNameWithPrefix
}

const Option = ({
  children,
  isSelected,
  isFocused,
  isDisabled,
  innerProps,
  innerRef,
  labelComponent,
  cx,
  withCheckbox
}) => (
  <div
    {...innerProps}
    ref={innerRef}
    className={withPrefix(
      cx,
      classNames(styles['select-option'], {
        [styles['select-option--selected']]: isSelected && !withCheckbox,
        [styles['select-option--focused']]: isFocused,
        [styles['select-option--disabled']]: isDisabled
      })
    )}
  >
    {withCheckbox && (
      <input
        type="checkbox"
        readOnly
        checked={isSelected}
        className={withPrefix(cx, styles['select-option__checkbox'])}
      />
    )}
    <span className={withPrefix(cx, styles['select-option__label'])}>
      <span className={withPrefix(cx, 'u-ellipsis')}>
        {labelComponent ? labelComponent : children}
      </span>
      {labelComponent ? children : false}
    </span>
    {!withCheckbox && (
      <span className={withPrefix(cx, styles['select-option__checkmark'])}>
        {isSelected && (
          <Icon
            icon="check-circleless"
            color={dodgerBlue}
            className={withPrefix(cx, 'u-ph-half')}
          />
        )}
      </span>
    )}
  </div>
)

Option.propTypes = {
  withCheckbox: PropTypes.bool,
  labelComponent: PropTypes.node
}

Option.defaultProps = {
  withCheckbox: false
}

const CheckboxOption = ({ ...props }) => <Option {...props} withCheckbox />

CheckboxOption.propTypes = {}

const ActionsOption = ({ actions, ...props }) => (
  <Option {...props} labelComponent={props.children}>
    <span className={withPrefix(props.cx, styles['select-option__actions'])}>
      {actions.map((action, index) => (
        <Icon
          key={index}
          icon={action.icon}
          color={props.isFocused ? coolGrey : silver}
          className={withPrefix(props.cx, 'u-ph-half')}
          onClick={e => {
            e.stopPropagation()
            action.onClick(props)
          }}
        />
      ))}
    </span>
  </Option>
)

ActionsOption.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      icon: PropTypes.string,
      onClick: PropTypes.func
    })
  )
}

ActionsOption.defaultProps = {
  actions: []
}

// With React, the referenced element is in the current property.
// With Preact, the referenced element is the object
const getNodeFromRef = ref => {
  return (ref && ref.current) || ref
}

const customComponents = {
  DropdownIndicator,
  Option
}

class SelectBox extends Component {
  constructor(props) {
    super(props)
    this.state = { isOpen: false }
    this.element = null
    this.handleRef = this.handleRef.bind(this)
    this.handleOpen = this.handleOpen.bind(this)
    this.handleClose = this.handleClose.bind(this)
  }

  handleOpen() {
    this.refreshMenuStyle()
    this.setState({
      isOpen: true
    })
  }

  refreshMenuStyle() {
    if (this.props.container && this.controlRef) {
      this.menuStyle = computedMenuListHeightStyles(
        getNodeFromRef(this.props.container),
        this.controlRef
      )
    }
  }

  handleClose() {
    this.setState({ isOpen: false })
  }

  handleRef(ref) {
    const { inputRef } = this.props
    if (ref && ref.select) {
      if (ref.select.controlRef) {
        // Save control ref to use for menu height computation
        this.controlRef = ref.select.controlRef
      }

      if (ref.select.inputRef && typeof inputRef === 'function') {
        inputRef(ref.select.inputRef)
      }
    }
  }

  render() {
    const {
      className,
      components,
      disabled,
      fullwidth,
      styles: reactSelectStyles,
      breakpoints: { isMobile },
      classNamePrefix,
      name,
      ...props
    } = this.props
    const showOverlay = this.state.isOpen && isMobile
    return (
      <ReactSelect
        ref={this.handleRef}
        components={
          components ? { ...customComponents, ...components } : customComponents
        }
        styles={{
          ...customStyles(this.props),
          ...reactSelectStyles,
          ...(this.menuStyle || {})
        }}
        onMenuOpen={this.handleOpen}
        onMenuClose={this.handleClose}
        {...props}
        isDisabled={disabled}
        className={classNames(
          {
            [styles['select__overlay']]: showOverlay,
            [styles['select--autowidth']]: !fullwidth,
            [styles['select--disabled']]: disabled,
            [styles['select--fullwidth']]: fullwidth
          },
          className
        )}
        // react-select temporarily adds className to its innerComponents
        // but this behavior will soon be removed. For the moment, we
        // cancel it by setting it to empty string
        // needsclick is added only on iOS App since fastclik tries to handle
        // the click and catch the event.
        classNamePrefix={
          isIOSApp()
            ? `needsclick ${classNamePrefix ? classNamePrefix : ''}`
            : classNamePrefix
            ? classNamePrefix
            : ''
        }
        selectProps={{ name }}
      />
    )
  }
}

SelectBox.propTypes = {
  container: PropTypes.object,
  components: PropTypes.object,
  disabled: PropTypes.bool,
  fullwidth: PropTypes.bool,
  name: PropTypes.string,
  inputRef: PropTypes.func,
  size: PropTypes.oneOf(['tiny', 'medium', 'large']),
  styles: PropTypes.object
}

SelectBox.defaultProps = {
  components: {},
  fullwidth: false,
  size: 'large',
  styles: {}
}

export default withBreakpoints()(SelectBox)
export {
  Option,
  CheckboxOption,
  ActionsOption,
  computedMenuListHeightStyles,
  reactSelectControl,
  components,
  SelectBox
}
