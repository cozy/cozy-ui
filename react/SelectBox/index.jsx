import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import ReactSelect from 'react-select'
import styles from './styles.styl'
import Icon from '../Icon'
import palette from '../../stylus/settings/palette.json'

const customStyles = {
  container: base => ({
    ...base,
    maxWidth: '30rem'
  }),
  control: (base, state) => ({
    ...base,
    backgroundColor: 'white',
    border: state.isFocused
      ? `.0625rem solid ${palette['dodgerBlue']}`
      : `.0625rem solid ${palette['silver']}`,
    ':hover': {
      borderColor: palette['coolGrey']
    },
    borderRadius: '.1875rem',
    boxShadow: 'unset',
    padding: '.503rem .5rem'
  }),
  dropdownIndicator: (base, state) => ({
    ...base,
    backgroundImage: state.menuIsOpen
      ? 'url("../../assets/icons/ui/icon-top-select.svg")'
      : 'url("../../assets/icons/ui/icon-bottom-select.svg")',
    backgroundSize: '.875rem',
    height: '.875rem',
    marginRight: '.75rem',
    padding: 0,
    width: '.875rem'
  }),
  indicatorSeparator: () => ({
    display: 'none'
  }),
  valueContainer: base => ({
    ...base,
    color: 'black'
  }),
  menu: base => ({
    ...base,
    zIndex: 10
  })
}

const reactSelectControl = CustomControl => ({ innerProps, children }) => (
  <div {...innerProps}>
    {CustomControl}
    <div className={styles['select-control__input']}>{children}</div>
  </div>
)

const Option = ({
  children,
  isSelected,
  isFocused,
  innerProps,
  withCheckbox
}) => (
  <div
    {...innerProps}
    className={classNames(styles['select-option'], {
      [styles['select-option--selected']]: isSelected && !withCheckbox,
      [styles['select-option--focused']]: isFocused
    })}
  >
    {withCheckbox && (
      <input
        type="checkbox"
        checked={isSelected}
        className={styles['select-option__checkbox']}
      />
    )}
    {children}
    {isSelected &&
      !withCheckbox && (
        <Icon
          icon="check-circleless"
          color={palette['dodgerBlue']}
          className={styles['select-option--checkmark']}
        />
      )}
  </div>
)

Option.propTypes = {
  withCheckbox: PropTypes.bool
}

Option.defaultProps = {
  withCheckbox: false
}

const CheckboxOption = ({ ...props }) => <Option {...props} withCheckbox />

CheckboxOption.propTypes = {}

const SelectBox = ({ components, styles, ...props }) => (
  <ReactSelect
    components={{ Option, ...components }}
    styles={{ ...customStyles, ...styles }}
    {...props}
  />
)

SelectBox.propTypes = {
  components: PropTypes.object,
  styles: PropTypes.object
}

SelectBox.defaultProps = {
  components: {},
  styles: {}
}

export * from 'react-select'
export default SelectBox
export { Option, CheckboxOption, reactSelectControl }
