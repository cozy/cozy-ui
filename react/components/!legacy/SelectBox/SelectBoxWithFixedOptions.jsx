import { components } from 'react-select'
import React from 'react'
import groupBy from 'lodash/groupBy'

import SelectBox from './SelectBox'
import styles from './styles.styl'

const Group = props => {
  return (
    <components.Group
      {...props}
      className={props.data.isFixed ? styles.FixedGroup : styles.Group}
    />
  )
}

const MenuList = props => (
  <components.MenuList {...props} className={styles.MenuList} />
)

const Nothing = () => null

const groupByFixed = options => {
  const { fixed = [], normal = [] } = groupBy(options, o =>
    o.fixed ? 'fixed' : 'normal'
  )
  return [
    { label: 'normal', options: normal },
    { label: 'fixed', isFixed: true, options: fixed }
  ]
}

const SelectBoxWithFixedOptions = props => (
  <SelectBox
    {...props}
    options={groupByFixed(props.options)}
    styles={{
      ...props.styles
    }}
    components={{
      MenuList: MenuList,
      GroupHeading: Nothing,
      Group,
      ...props.components
    }}
  />
)

SelectBoxWithFixedOptions.defaultProps = {
  styles: {},
  components: {}
}

export default SelectBoxWithFixedOptions
