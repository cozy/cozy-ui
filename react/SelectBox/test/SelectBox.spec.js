'use strict'
/* eslint-env jest */
import React from 'react'

import { computedMenuListHeightStyles } from '../SelectBox'
import { SelectBox } from '../SelectBox'
import { isIOSApp } from 'cozy-device-helper'

jest.mock('cozy-device-helper', () => ({
  ...require.requireActual('cozy-device-helper'),
  isIOSApp: jest.fn()
}))

describe('SelectBox', () => {
  describe('computedMenuListHeightStyles', () => {
    const fixtures = {
      containerMock: {
        getBoundingClientRect: jest.fn().mockReturnValue({
          bottom: 235
        }),
        style: {
          paddingTop: '1rem',
          paddingBottom: '2rem'
        }
      },
      selectMock: {
        getBoundingClientRect: jest.fn().mockReturnValue({
          bottom: 100
        })
      },
      stylesBaseMock: {
        paddingTop: 2,
        paddingBottom: 2
      }
    }

    it('returns expected style object', () => {
      const styleClosures = computedMenuListHeightStyles(
        fixtures.containerMock,
        fixtures.selectMock
      )
      expect(typeof styleClosures.menuList).toBe('function')

      expect(styleClosures.menuList(fixtures.stylesBaseMock)).toEqual({
        paddingTop: 2,
        paddingBottom: 2,
        maxHeight: 'calc(131px - 1rem - 2rem)'
      })
    })

    it('returns expected style object when container has no padding', () => {
      const styleClosures = computedMenuListHeightStyles(
        { ...fixtures.containerMock, style: {} },
        fixtures.selectMock
      )
      expect(typeof styleClosures.menuList).toBe('function')

      expect(styleClosures.menuList(fixtures.stylesBaseMock)).toEqual({
        paddingTop: 2,
        paddingBottom: 2,
        maxHeight: 'calc(131px - 0px - 0px)'
      })
    })

    it('returns expected style object when base styles have no padding', () => {
      const styleClosures = computedMenuListHeightStyles(
        fixtures.containerMock,
        fixtures.selectMock
      )
      expect(typeof styleClosures.menuList).toBe('function')

      expect(styleClosures.menuList({})).toEqual({
        maxHeight: 'calc(135px - 1rem - 2rem)'
      })
    })
  })
  describe('needsclick', () => {
    beforeEach(() => {
      jest.resetModules()
      isIOSApp.mockReturnValue(false)
    })
    it('should not add needclicks if not on iOS', () => {
      const wrapper = shallow(<SelectBox breakpoints={{}} />)
      expect(wrapper.prop('classNamePrefix')).toEqual('')
    })
    it('shoudl add needclick if iOS', () => {
      isIOSApp.mockReturnValue(true)
      const wrapper = shallow(<SelectBox breakpoints={{}} />)
      expect(wrapper.prop('classNamePrefix')).toContain('needsclick')
    })
  })
})
