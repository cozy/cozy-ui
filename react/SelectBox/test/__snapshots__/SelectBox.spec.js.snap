// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBox needsclick shoudl add needclick if iOS 1`] = `
<StateManager
  className="styles__select--autowidth___16AEp"
  classNamePrefix="needsclick "
  components={
    Object {
      "DropdownIndicator": [Function],
      "Option": [Function],
    }
  }
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  onMenuClose={[Function]}
  onMenuOpen={[Function]}
  selectProps={
    Object {
      "name": undefined,
    }
  }
  size="large"
  styles={
    Object {
      "control": [Function],
      "dropdownIndicator": [Function],
      "indicatorSeparator": [Function],
      "menu": [Function],
      "valueContainer": [Function],
    }
  }
/>
`;

exports[`SelectBox needsclick should not add needclicks if not on iOS 1`] = `
<StateManager
  className="styles__select--autowidth___16AEp"
  classNamePrefix=""
  components={
    Object {
      "DropdownIndicator": [Function],
      "Option": [Function],
    }
  }
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  onMenuClose={[Function]}
  onMenuOpen={[Function]}
  selectProps={
    Object {
      "name": undefined,
    }
  }
  size="large"
  styles={
    Object {
      "control": [Function],
      "dropdownIndicator": [Function],
      "indicatorSeparator": [Function],
      "menu": [Function],
      "valueContainer": [Function],
    }
  }
/>
`;
